tools/mutation_test/mutator.py

```diff
@@ -0,0 +1,499 @@
+            else:
+                self.lines_to_mutate[old] = _flatten(_merge(self.src[f]))
+        self.rng = rng
+        self.mutation_patterns = self.flattenPatterns(mutation_patterns)
+        self.pattern_generator = self.mutation_patterns.copy()
+        self.file_index = None
+
+    def restore(self):
+        """ Restores source files by replacing src with old
+
+        Path should be root directory /freertos
+        """
+        for i in range(len(self.src)):
+            shutil.copyfile(self.olds[i], self.modified[i])
+
+    def cleanup(self):
+        """ Cleans up by deleting all old files
+        """
+        self.restore()
+        for i in range(len(self.src)):
+            os.remove(self.olds[i])
+
+    def mutate(self, occurrence: Occurrence) -> Tuple[str, str]:
+        """ Mutates given `occurrence` using `occurrence.pattern`
+
+        Returns the original line and the mutated line
+
+        See mutator.py for information on what a Occurrence object is
+        """
+        mutation_pattern = occurrence.pattern
+        source_code = open(occurrence.file).read().split('\n')
+        sys.stderr.write("\n==> @ Line: " + str(occurrence.line) + "\n\n")
+        sys.stderr.write("Original Line  : " + source_code[occurrence.line - 1].strip() + "\n")
+        mutated_line = (source_code[occurrence.line - 1][0:occurrence.index] + 
+            source_code[occurrence.line - 1][occurrence.index:]
+                .replace(mutation_pattern.pattern, mutation_pattern.transformation, 1))
+        sys.stderr.write("After Mutation : " + mutated_line.strip() + "\n")
+        self.write_to_file(occurrence.file.rstrip('.old'), source_code, occurrence.line - 1, mutated_line)
+        return source_code[occurrence.line - 1], mutated_line
+    
+    def write_to_file(self, path, source_code, mutated_line_number, mutated_line):
+        """ Writes a file to `path` by replacing line at `mutated_line_number` in `source_code`
+        with `mutated_line`.
+
+        Helper to separate write to file process with mutate function.
+        """
+        output_file = open(path, "w")
+        for i in range(0,len(source_code)) :
+            if i == mutated_line_number : 
+                output_file.write("/* XXX: original code was : " + source_code[i] + " */\n")
+                output_file.write(mutated_line + "\n")
+            else :
+                output_file.write(source_code[i] + "\n")
+        output_file.close()
+        sys.stderr.write("\nOutput written to " + path + "\n\n")
+
+    def findOccurrences(self, mutation_pattern: Pattern) -> List[Occurrence]:
+        """ Finds all occurrences of given `mutation_pattern` in current Mutator
+
+        Returns a list of Occurrence objects (see mutator.Occurrence in mutator.py)
+        
+        If there are multiple source files, searches all of them. Occurrence object
+        contains fields to access file, line, and index within line to find substring.
```

Comment:
can you also document how index is choosen.

Reply:
https://github.com/aws/amazon-freertos/pull/2475/commits/b00aeead808cb9ffbf76c5f0fae88e2d3ed8a223